Design.md

```diff
@@ -0,0 +1,187 @@
+| client_key.pem | the client's private key. |
+| root_ca.pem | the certificate of the CA that can verify the server's certificate. |
+
+Use CloudFlare's `cfssl` CLI to create and sign all of the certs mentioned above.
+
+#### Creating the self-signed root CA
+`cfssl selfsign -config cfssl.json --profile rootca "Greenbeans Dev CA" csr.json | cfssljson -bare root`
+
+[a relative link](cfssl.json)
+[a relative link](csr.json)
+
+# Creating and signing the server certificate:
+```
+cfssl genkey csr.json | cfssljson -bare server
+cfssl sign -ca root.pem -ca-key root-key.pem -config cfssl.json -profile server server.csr | cfssljson -bare server
+
+```
+
+# Creating and signing the client(s) certificates:
+```
+cfssl genkey csr.json | cfssljson -bare client
+cfssl sign -ca root.pem -ca-key root-key.pem -config cfssl.json -profile client client.csr | cfssljson -bare client
+```
+Multiple clients are supported, but each client will require it own certificate. The calls above can easily be modified to include additional client information (name):
+```
+cfssl genkey csr.json | cfssljson -bare client_alex
+cfssl sign -ca root.pem -ca-key root-key.pem -config cfssl.json -profile client client_alex.csr | cfssljson -bare client_alex
+```
+The CLI will include an optional parameter to specify the `client` key name, otherwise will default to 'client'.
+
+
+### Privacy
+Users can only interact with jobs that they started.
+
+### UX - CLI
+Scenarios<br />
+Start a job:<br />
+caller provides the command and arguments in a single string *** explore options 
+`jws start -c "ls -l"`<br />
+returns the unique identifier of the job 
+
+Stop a job:<br />
+`jws stop -j <job_id>`<br />
+returns an error if something went wrong (unknown user, invalid job_id, access denied)
+
+Get the status of a job:<br />
+`jws status -j <job_id>`<br />
+returns the status of the job: "Pending, Active, Terminated"
+  or an error: (uknown user, invalid job_id, access denied)
+
+Get the output of a job:<br />
+`jws output -j <job_id>`<br />
+
+
+### Proto Specifications
+```
+service JobWorkerService {
+  rpc StartJob(StartJobRequest) returns (StartJobResponse);
+  rpc StopJob(StopJobRequest) returns (google.protobuf.empty);
+  rpc GetJobStatus(GetJobStatusRequest) returns (GetJobStatusResponse);
+  rpc GetJobOutput(GetJobOutputRequest) returns (stream GetJobOutputResponse);
+}
+
+message StartJobRequest {
+  string cmd = 1;
+}
+
+message StartJobResponse {
+  string job_id = 1;
+}
+
+message StopJobRequest {
+  string job_id = 1;
+}
+
+message GetJobStatusRequest {
+  string job_id = 1;
+}
+
+enum JobStatus {
+  JOB_STATUS_UNDEFINED = 0;
+  JOB_STATUS_PENDING = 1;
+  JOB_STATUS_ACTIVE = 2;
+  JOB_STATUS_USER_CANCELLED = 3;
+  JOB_STATUS_TERMINATEDS = 4
+}
+
+message GetJobStatusResponse {
+  JobStatus status = 1;
+}
+
+message GetJobOutputRequest {
+  string job_id = 1;
+}
+
+message GetJobOutputResponse {
+  oneof output_msg {
+    bytes stdout = 1;
+    string stderr = 2;
```

Comment:
why string for stderr?

Reply:
will switch to bytes (original thought being that stderr was human readable and is a string,)

Reply:
actually switched back to just a single output. going to redirect stderr to stdout (cutting corners)